osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs

```diff
@@ -20,20 +23,116 @@ public class Speed : Skill
         private const double pi_over_4 = Math.PI / 4;
         private const double pi_over_2 = Math.PI / 2;
 
+        private const double streamaimconst = 2.42;
+
+        //  private const double stdevconst = 0.149820;
+
+        private int sdsplitcounter = 0;
+        private int sdsplitcounter2 = 0;
+        // private int runcheck1 = 0;
+
+        private double sdstrainmult = 0;
+        // private double sdstrainmult2;
+
+        // private double sdstrainmult2;
+
+        public readonly List<double> JumpDistances = new List<double>();
+        public readonly List<double> StrainTimes = new List<double>();
+        public List<double> JumpDistances2 = new List<double>();
+
         protected override double SkillMultiplier => 1400;
         protected override double StrainDecayBase => 0.3;
 
         private const double min_speed_bonus = 75; // ~200BPM
         private const double max_speed_bonus = 45; // ~330BPM
         private const double speed_balancing_factor = 40;
 
+        private double calculateStandardDeviation(IEnumerable<double> values)
+        {
+            double standardDeviation = 0;
+
+            if (values.Any())
+            {
+                // Compute the average.     
+                double avg = values.Average();
+
+                // Perform the Sum of (value-avg)_2_2.      
+                double sum = values.Sum(d => Math.Pow(d - avg, 2));
+
+                // Put it all together.      
+                standardDeviation = Math.Sqrt((sum) / (values.Count()));
+            }
+
+            return standardDeviation;
+        }
         protected override double StrainValueOf(DifficultyHitObject current)
         {
             if (current.BaseObject is Spinner)
                 return 0;
 
             var osuCurrent = (OsuDifficultyHitObject)current;
 
+            double sectionvelocity = osuCurrent.JumpDistance / osuCurrent.StrainTime;
+
+            if (osuCurrent.JumpDistance < 150 && sectionvelocity < streamaimconst && sectionvelocity > 0.9 && Previous.Count > 0 && osuCurrent.Angle != null && osuCurrent.Angle.Value >= Math.PI / 2 && osuCurrent.StrainTime < 100)
```

Comment:
This conditional is way too complex to remain in its current state.

Reply:
Can you suggest another way to condition that?

Reply:
I don't really even have the slightest clue what this is *attempting* to do without a proper explanation, so not currently, no. Factoring out a separate method might be good to start with.

Reply:
It is designed to seperate jumps from streams and sectionvelocity > 0.9 part is for seperating streams into really hard streams for further calculation.

Reply:
So many questions here. *How* does it separate? What's the methodology here? Why were the thresholds chosen the way they were?

In other words, how is anyone supposed to change this code in a month or a year, without reverse engineering it all themselves?

Reply:
Do you want a white paper like Bitcoin White Paper?

Reply:
- osuCurrent.JumpDistance < 150 => Excors 9* Last stream

- sectionvelocity < streamaimconst => Excors 9* Last stream

- sectionvelocity > 0.9 => Packet Hero [Ruthless] nerf on slow acceleration changing stream

- Previous.Count > 0 && osuCurrent.Angle != null => Initial checks if angle measurement possible or not?

- osuCurrent.Angle.Value >= Math.PI / 2 => Streams must have angle greater that 90 degrees. ( Suggested by abraker )

- osuCurrent.StrainTime < 100 ( Equivalent to 150 bpm streams-ish (Not exact due to Beat Snap Divisor ) => Consensus threshold for stream and jump seperation

The full methodology and my work history is in Discord.

Reply:
Yeah this explanation makes me opposed to this being merged at all ever. It's mostly based on seemingly random examples to nerf/buff specific maps as you see fit.